### PR TITLE
Fix project config based run with github-[pr-]check and github-pr-review reporters

### PR DESCRIPTION
**Pull request by haya14busa**
- Tool name was incorrect for github-[pr-]check reporters and check API does not work correctly for project config based run (ref: https://github.com/reviewdog/reviewdog/issues/1846)
- Deletion of outdated comment didn't work properly with github-pr-review reporter since tool name isn't correctly configured for project config based run

- [x] Updated Unreleased section in [CHANGELOG](https://github.com/reviewdog/reviewdog/blob/master/CHANGELOG.md) or it's not notable changes.

### CODE REVIEW

File: service/github/check.go [hunk 1]
```diff
@@ -78,9 +78,14 @@ func (ch *Check) GetResult() *CheckResult {
 	return ch.result
 }
 
+func (ch *Check) SetToolName(toolName string) {
```
Comment (review-dog) [inline service/github/check.go:81]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.SetToolName should have comment or be unexported

File: service/github/check.go [hunk 2]
```diff
@@ -78,9 +78,15 @@ func (ch *Check) GetResult() *CheckResult {
 	return ch.result
 }
 
+func (ch *Check) SetTool(toolName string, level string) {
```
Comment (review-dog) [inline service/github/check.go:81]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.SetTool should have comment or be unexported

File: service/github/github.go [hunk 3]
```diff
@@ -123,6 +124,11 @@ func (g *PullRequest) Flush(ctx context.Context) error {
 	return g.postAsReviewComment(ctx)
 }
 
+func (g *PullRequest) SetTool(toolName string, level string) {
```
Comment (review-dog) [inline service/github/github.go:127]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method PullRequest.SetTool should have comment or be unexported

File: service/bitbucket/annotator.go [hunk 4]
```diff
@@ -99,6 +99,7 @@ func (r *ReportAnnotator) Post(_ context.Context, c *reviewdog.Comment) error {
 // Flush posts comments which has not been posted yet.
 func (r *ReportAnnotator) Flush(ctx context.Context) error {
 	r.muAnnotations.Lock()
+	defer func() { r.comments = nil }()
 	defer r.muAnnotations.Unlock()
```
Comment (shogo82148) [inline service/bitbucket/annotator.go:103]: The defer statement is executed in LIFO order.
In this case, the order is `r.muAnnotations.Unlock()`, `r.comments = nil`.
Is it intentional that `r.comments = nil` is outside the lock scope?

```suggestion
	r.muAnnotations.Lock()
	defer r.muAnnotations.Unlock()
	defer func() { r.comments = nil }()
```

Comment (haya14busa) [reply service/bitbucket/annotator.go:103]: Oh, may bad. Thank you for the review!

File: service/gerrit/change_review.go [hunk 5]
```diff
@@ -60,6 +60,7 @@ func (g *ChangeReviewCommenter) Post(_ context.Context, c *reviewdog.Comment) er
 // Flush posts comments which has not been posted yet.
 func (g *ChangeReviewCommenter) Flush(ctx context.Context) error {
 	g.muComments.Lock()
+	defer func() { g.postComments = nil }()
 	defer g.muComments.Unlock()
```
Comment (shogo82148) [inline service/gerrit/change_review.go:64]: same as https://github.com/reviewdog/reviewdog/pull/1867#discussion_r1703101026

```suggestion
	g.muComments.Lock()
	defer g.muComments.Unlock()
	defer func() { g.postComments = nil }()
```

File: service/gitea/gitea.go [hunk 6]
```diff
@@ -91,6 +91,7 @@ func (g *PullRequest) Post(_ context.Context, c *reviewdog.Comment) error {
 // Flush posts comments which has not been posted yet.
 func (g *PullRequest) Flush(_ context.Context) error {
 	g.muComments.Lock()
+	defer func() { g.postComments = nil }()
 	defer g.muComments.Unlock()
```
Comment (shogo82148) [inline service/gitea/gitea.go:95]:
```suggestion
	g.muComments.Lock()
	defer g.muComments.Unlock()
	defer func() { g.postComments = nil }()
```

File: service/github/github.go [hunk 7]
```diff
@@ -115,6 +115,7 @@ func (g *PullRequest) Post(_ context.Context, c *reviewdog.Comment) error {
 // Flush posts comments which has not been posted yet.
 func (g *PullRequest) Flush(ctx context.Context) error {
 	g.muComments.Lock()
+	defer func() { g.postComments = nil }()
 	defer g.muComments.Unlock()
```
Comment (shogo82148) [inline service/github/github.go:119]:
```suggestion
	g.muComments.Lock()
	defer g.muComments.Unlock()
	defer func() { g.postComments = nil }()
```

File: service/gitlab/gitlab_mr_commit.go [hunk 8]
```diff
@@ -69,6 +69,7 @@ func (g *MergeRequestCommitCommenter) Post(_ context.Context, c *reviewdog.Comme
 // Flush posts comments which has not been posted yet.
 func (g *MergeRequestCommitCommenter) Flush(ctx context.Context) error {
 	g.muComments.Lock()
+	defer func() { g.postComments = nil }()
 	defer g.muComments.Unlock()
```
Comment (shogo82148) [inline service/gitlab/gitlab_mr_commit.go:73]:
```suggestion
	g.muComments.Lock()
	defer g.muComments.Unlock()
	defer func() { g.postComments = nil }()
```

File: service/gitlab/gitlab_mr_discussion.go [hunk 9]
```diff
@@ -71,6 +71,7 @@ func (g *MergeRequestDiscussionCommenter) Post(_ context.Context, c *reviewdog.C
 // Flush posts comments which has not been posted yet.
 func (g *MergeRequestDiscussionCommenter) Flush(ctx context.Context) error {
 	g.muComments.Lock()
+	defer func() { g.postComments = nil }()
 	defer g.muComments.Unlock()
```
Comment (shogo82148) [inline service/gitlab/gitlab_mr_discussion.go:75]:
```suggestion
	g.muComments.Lock()
	defer g.muComments.Unlock()
	defer func() { g.postComments = nil }()
```

**Comment by haya14busa** [comment]
Oops, thank you for the suggestion! PTAL.

**Review by shogo82148 — APPROVED** [review]
LGTM!

**Pull request merged**